### ISSUE TITLE
Topological parallel grouping for script execution in workspaces

### ISSUE THREAD

**Issue by millsp** [pnpm/pnpm#7630]
### Contribution

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

Using `--parallel` can drastically speed up build scripts. However, when the ordering of the build commands is important, `--parallel` can obviously not work, and we must resort to only using `-r`. Running topological groups in parallel could reduce the time a build step could take. To illustrate this, let's take a look at the Prisma monorepo packages build commands when run with `pnpm -r dev`:

```ts
// each of these is run one after another
packages/debug build$ tsx helpers/build.ts
packages/driver-adapter-utils build$ tsx helpers/build.ts
packages/adapter-d1 build$ tsx helpers/build.ts
packages/adapter-libsql build$ tsx helpers/build.ts
packages/adapter-neon build$ tsx helpers/build.ts
packages/adapter-pg build$ tsx helpers/build.ts
packages/adapter-planetscale build$ tsx helpers/build.ts
packages/generator-helper build$ tsx helpers/build.ts
packages/get-platform build$ tsx helpers/build.ts
packages/fetch-engine build$ tsx helpers/build.ts
packages/engines build$ tsx helpers/build.ts
packages/internals build$ tsx helpers/build.ts
packages/instrumentation build$ tsx helpers/build.ts
packages/migrate build$ tsx helpers/build.ts
packages/client build$ tsx helpers/build.ts
packages/integration-tests build$ tsx helpers/build.ts
packages/cli build$ tsx helpers/build.ts
```

If we had topological group parallelism (not sure if this is the right term, but I will illustrate it), we could then batch some builds in parallel while respecting the topological order. Here's how it would look like:

```ts
// all packages need these two, they run in parallel in their own topo group
packages/debug build$ tsx helpers/build.ts
packages/driver-adapter-utils build$ tsx helpers/build.ts

// these packages need the previous block, they run in parallel in their own topo group
packages/adapter-d1 build$ tsx helpers/build.ts
packages/adapter-libsql build$ tsx helpers/build.ts
packages/adapter-neon build$ tsx helpers/build.ts
packages/adapter-pg build$ tsx helpers/build.ts
packages/adapter-planetscale build$ tsx helpers/build.ts
packages/generator-helper build$ tsx helpers/build.ts

// this package is needed by other packages below, runs in it's own top group
packages/get-platform build$ tsx helpers/build.ts

// these packages need the previous block, they run in their own topo group
packages/fetch-engine build$ tsx helpers/build.ts

// these packages need the previous block, they run in their own topo group
packages/engines build$ tsx helpers/build.ts

// these packages need the previous block, they run in parallel in their own topo group
packages/internals build$ tsx helpers/build.ts
packages/instrumentation build$ tsx helpers/build.ts
packages/migrate build$ tsx helpers/build.ts
packages/client build$ tsx helpers/build.ts
packages/integration-tests build$ tsx helpers/build.ts

// these packages need the previous block, they run in their own topo group
packages/cli build$ tsx helpers/build.ts
```

So basically, each of the items in these groups can be run in parallel, but each group needs to be serial.

### Describe the solution you'd like

Please correct me if I am wrong, but I haven't found a way to do this (besides hardcoding it or writing a script). I'd like an option like `--group, -g` or `--topo, -t` to go along with `-r --parallel`. I think this could provide considerable build time gains (and perhaps other scripts too) on huge monorepos.

### Describe the drawbacks of your solution

I don't think there is any, it's purely additional and does not change the current behavior.

### Describe alternatives you've considered

Hardcoding or writing a script.

**Comment by zkochan** [comment]
It should already work as you described, when you use `pnpm -r`. Unless, you use `pnpm -r --no-sort` (https://pnpm.io/cli/recursive#--no-sort)

**Comment by GerkinDev** [comment]
There might be a regression here: with v8.3.1, topology of workspace is respected. With 8.15.3, it is not.

Last working version is 8.6.12. First breaking one is 8.7.0.

After digging a bit more, I spotted that it might be due to the resolution of https://github.com/pnpm/pnpm/issues/6922; in my case, replacing relative links to workspace packages with `workspace:*` worked. Yet, this is IMO a breaking change that should be mentioned, and the section [in the docs](https://pnpm.io/workspaces#referencing-workspace-packages-through-their-relative-path) using relative path might be out of date

**Comment by ahaoboy** [comment]
Good question, consider the following simple case

```
a(1) - b(10) - d(1)
  \      \
   c(10)  e(1)
```

The result of topological sorting is
```
[ 'c', 'd', 'e' ], [ 'b' ], [ 'a' ]
```
When pnpm executes, it takes one group as a whole, so cde takes 10s, b takes 10s, a takes 1s, and the total takes 21s.

```
 time pnpm -r build

Executed in   21.24 secs      fish           external
```

But we can clearly see: when de finishes executing, we can start b's execution, de takes 1s, cb takes 10s, a takes 1s, and it only takes 12 seconds.

Most tools don't implement this optimization, and it's not hard to implement, it's just that pnpm's execution logic is too complex.

Using the --no-sort parameter disrupts the order of dependencies
```
 time pnpm -r --no-sort build
Scope: 5 of 6 workspace projects
c build$ sleep 10 && echo c
│ c
└─ Done in 10s
d build$ sleep 1 && echo d
│ d
└─ Done in 1s
a build$ sleep 1 && echo a
│ a
└─ Done in 1s
b build$ sleep 10 && echo b
│ b
└─ Done in 10s
e build$ sleep 1 && echo e
│ e
└─ Done in 1s
```
demo: 
https://github.com/ahaoboy/pnpm-graph-demo

**Comment by RexSkz** [comment]
Since pnpm uses `p-limit`, which can only restrict the max concurrent tasks but can't read their dependencies, "grouping tasks by dependency and executing each group in sequence" seems to be the best choice.

If pnpm wants to make use of each worker, it must implement a data structure like this:

```ts
interface Task {
  id: string;        // or any other values that can be used as "key"
  fn: Promise<void>;
  prevIds: string[]; // to judge whether this task can be run
  nextIds: string[]; // to trigger next tasks
}
```

Then it can flatten the "groups" (each cycle should be a subsequence of it) and use it to create a `Task[]`, for example:

```ts
[
  { id: 'a', fn: ..., prevIds: [], nextIds: ['c', 'e'] },
  { id: 'b', fn: ..., prevIds: [], nextIds: ['d'] },
  { id: 'c', fn: ..., prevIds: ['a'], nextIds: ['e'] },
  { id: 'd', fn: ..., prevIds: ['b'], nextIds: [] },
  { id: 'e', fn: ..., prevIds: ['a', 'c'], nextIds: [] },
]
```

Finally, we can implement a task runner with concurrency limit:

- When a task `x` is triggered, it uses `x.prevIds` to judge whether it can be run.
- When a task `x` is finished, it tries to trigger all tasks whose `id` is in `x.nextIds`.

**Comment by ahaoboy** [comment]
>  "grouping tasks by dependency and executing each group in sequence" seems to be the best choice. 
 
Maybe we can think a little further ahead, we can borrow ideas like [parcel-graph](https://www.npmjs.com/package/@parcel/graph), performance and memory will be better than Map, and topological sorting, group execution, interrupt and resume execution of these features should be able to be reused by other package management

**Comment by aderant-smcd** [comment]
I created a dedicated issue for the grouping inefficiency with a simple repro here: https://github.com/pnpm/pnpm/issues/9885

**Comment by zkochan** [comment]
This is implemented in current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm -r run` runs projects that do not depend on each other at the same time and waits only where there is a dependency. Three 1.5 s scripts with one dependency edge finished in 3.1 s.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
